### PR TITLE
docs: add visual testing tooling RFC

### PR DESCRIPTION
**Pull request by separatio**
[FX-2534]

### Description

This is the RFC for visual tests in Picasso.
Should be self-explanatory.

### How to test

- review the doc

### Screenshots

| Before.                                 | After.                                  |
| --------------------------------------- | --------------------------------------- |
| Insert screenshots or screen recordings | Insert screenshots or screen recordings |

### Development checks

- [x] Self reviewed

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-2534]: https://toptal-core.atlassian.net/browse/FX-2534?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/decisions/04-single-visual-tests-tool.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+1. Puppeteer with Storybook
+2. Happo with Cypress
+
+## Proposal
+
+Move all visual tests to Cypress with Happo.
+
+The reasons for this are:
+- Picasso will support only one provider (Happo) and framework (Cypress) for visual tests
+- PRs will have only one visual tests check
```
Comment (denieler) [inline docs/decisions/04-single-visual-tests-tool.md:15]: I don't think this will be true, because we probably would still want to distinguish Cypress tests from Storybook tests

File: docs/decisions/04-single-visual-tests-tool.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+- Storybook's purpose in Picasso is not for doing visual tests
+- supporting two different plugins (Storybook and Happo) brings additional complexity and maintenance effort
```
Comment (denieler) [inline docs/decisions/04-single-visual-tests-tool.md:17]: we probably need to say that maintaining `jest-image-snapshot` (with visual test result files pushed to git) and Happo at the same time is complex. Also, we need to check visual screenshots using 2 different tools, which is confusing

File: docs/decisions/04-single-visual-tests-tool.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+## Problem
+
+Two different frameworks for visual testing are currently supported in Picasso:
+1. Puppeteer with Storybook
+2. Happo with Cypress
+
+## Proposal
+
+Move all visual tests to Cypress with Happo.
```
Comment (denieler) [inline docs/decisions/04-single-visual-tests-tool.md:11]: Maybe removing completely Storybook tests would not play that well for us, because Storybook is one of the outcomes, and products, which we propose to our users. We would like to make sure that our products look great. But at the same time, we don't want to have 2 different technologies. Maye the better way to go would be to do visual tests of Storybook via Happo? WDYT?

File: docs/decisions/04-single-visual-tests-tool.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+- it will take some time to migrate tests to Cypress
+- manual intervention in adding visual tests later
+- the Puppeteer stack (and tests) can't be removed immediately
+
+## Alternatives
+
+1. Simply replace the Puppeteer stack with the Happo provider for visual tests in Storybook:
+- this is the easy way out, but gives us less control and visibility over what's covered
+- maintenance over Cypress visual tests is still needed, as they are covering different areas
+- it requires supporting and maintaining two configurations for visual tests
```
Comment (denieler) [inline docs/decisions/04-single-visual-tests-tool.md:36]: maybe we can merge those two configurations and use the same Happo engine on both types of tests 🤔

File: docs/decisions/04-single-visual-tests-tool.md (hunk 1 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md:15]: For the current proposal, there are no Storybook tests. :slightly_smiling_face: 
Of course, if we have both, we'll get two checks in PRs.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 3 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md:11]: I've considered it as well, hence the Alternatives section.
tl;dr It's going to be hard to take decisions what to test in which side if we do that.

These are some bad possible outcomes I am foreseeing:
- component is visually tested in both Storybook and Cypress, or at least they have overlapping areas being tested
- extending examples to increase visual test coverage instead of increasing test coverage via Cypress
- increased amount of time when deciding the "correct" way to visually test a component, bringing more discussions and opposing views, rather than unity

Yes, the product is showcased in Storybook. Hence, I am seeing Storybook as our Production environment. We shouldn't be testing on Production environments.

On the other hand, if we abandon Cypress visual tests, we'll end up with many potentially unnecessary examples in Storybook, bloating it and making it less usable for our end users.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 4 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md:36]: We need to configure two separate plugins regardless. To my knowledge, the Storybook plugin is too different from the Cypress plugin to merge the configs. Check these for more information:

https://docs.happo.io/docs/cypress
https://docs.happo.io/docs/storybook

File: docs/decisions/04-single-visual-tests-tool.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
+# Single Visual Tests tool
+
+## Problem
+
+Two different frameworks for visual testing are currently supported in Picasso:
```
Comment (ozgurkececioglu) [inline docs/decisions/04-single-visual-tests-tool.md:5]: I'm not quite sure what is the problem 🙂 I mean it could be more clear

File: docs/decisions/04-single-visual-tests-tool.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+- the Puppeteer stack (and tests) can't be removed immediately
+
+## Alternatives
+
+1. Simply replace the Puppeteer stack with the Happo provider for visual tests in Storybook:
+- this is the easy way out, but gives us less control and visibility over what's covered
+- maintenance over Cypress visual tests is still needed, as they are covering different areas
+- it requires supporting and maintaining two configurations for visual tests
+
+2. A mix between Storybook and Cypress visual tests, both using the Happo provider plugins:
```
Comment (TomasSlama) [inline docs/decisions/04-single-visual-tests-tool.md:38]: I don't see the difference between the first alternative and this one. 🤔,

When writing cypress tests many times it seems we just copy-paste examples from stories. So in those cases, it would be great if we could just take a screenshot in the storybook.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 3 above)
Comment (dmaklygin) [reply docs/decisions/04-single-visual-tests-tool.md:11]: This is an existing plugin (https://docs.happo.io/docs/storybook) to use Happo with Storybook.
* Cypress will test only behavioral tests, meanwhile visual tests in Storybook will take a screenshot of the component state
* As far as I know, we are not going to increase the number of tests. Instead, we are willing to have a standard approach to writing/maintaining visual tests: keeping only Storybook+Happo will let us achieve this.
* I think, this is the most simple task, we will take a screenshot of each story in the Storybook, as we currently do

File: docs/decisions/04-single-visual-tests-tool.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+Move all visual tests to Cypress with Happo.
+
+The reasons for this are:
+- Picasso will support only one provider (Happo) and framework (Cypress) for visual tests
+- PRs will have only one visual tests check
+- Storybook's purpose in Picasso is not for doing visual tests
+- supporting two different plugins (Storybook and Happo) brings additional complexity and maintenance effort
+
+Considering this move, we will gain:
+- control over which visual tests we add
```
Comment (dmaklygin) [inline docs/decisions/04-single-visual-tests-tool.md:20]: Hmm, honestly, I would say the same for every single technology.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 5 above)
Comment (OleksandrNechai) [reply docs/decisions/04-single-visual-tests-tool.md:5]: Yeah, I guess real problem is Puppeteer being super slow and hence we have slow feedback from quality gate which slows down development and reduce developer's focus.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 6 above)
Comment (OleksandrNechai) [reply docs/decisions/04-single-visual-tests-tool.md:38]: As I understand, the difference is that in first option there is no place for Cypress tests.

File: docs/decisions/04-single-visual-tests-tool.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+### Drawbacks and limitations
+
+There are two drawbacks present to the above proposal:
+- it will take some time to migrate tests to Cypress
+- manual intervention in adding visual tests later
+- the Puppeteer stack (and tests) can't be removed immediately
+
+## Alternatives
+
+1. Simply replace the Puppeteer stack with the Happo provider for visual tests in Storybook:
```
Comment (OleksandrNechai) [inline docs/decisions/04-single-visual-tests-tool.md:33]: To me alternative (1) where we do not use Cypress simply is not possible. Take Dropdown for example. We need to open it and make screenshot. It can only be done in Cypress (maybe also in Storybook with some scripts, but is not it abusing Storybook?). Next, we need to test components composition. For example, Dropdown on Modal. Only Cypress for this.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 6 above)
Comment (OleksandrNechai) [reply docs/decisions/04-single-visual-tests-tool.md:38]: Tomas makes good point. If we go with Cypress only, we will have many code duplicates in stories and tests. Not sure if this is a problem though.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 3 above)
Comment (OleksandrNechai) [reply docs/decisions/04-single-visual-tests-tool.md:11]: >I am seeing Storybook as our Production environment. We shouldn't be testing on Production environments.

Right. Yet we test employs. And it makes total sense to test if what we potentially deploy looks reasonable. What if all components are rendered without styles? We will not notice it.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 2 above)
Comment (OleksandrNechai) [reply docs/decisions/04-single-visual-tests-tool.md:17]: I see only 2 real options: Cypress vs Cypress+Storybook. Now, is this `complexity and maintenance effort` so huge that we should not test Storybook to avoid it? Pleas, help us understand what are these efforts?

File: docs/decisions/04-single-visual-tests-tool.md (hunk 7 above)
Comment (OleksandrNechai) [reply docs/decisions/04-single-visual-tests-tool.md:20]: I think the idea here is that we actively create Cypress tests instead of having all automatically covered in Storybook and tests kind appear against our active participation. Sounds like a good thing, we all learned hard way that `magic` feels good at first and then it is hard to maintain.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 5 above)
Comment (OndrejTuma) [reply docs/decisions/04-single-visual-tests-tool.md:5]: I agree, some summarization of headaches that we have with using these two frameworks together would be nice 🙂

Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md:5]: Is the problem clearly stated now @ozgurkececioglu @OleksandrNechai @OndrejTuma ? Is there anything that might be a good addition to this section?

File: docs/decisions/04-single-visual-tests-tool.md (hunk 3 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md:28]: I understand your concern, @LashaJini . However, it still remains true that some of us and contributors are having headaches due to permissions in Linux, giving us headaches in running the tests, not to mention update screenshots.

Furthermore, we've already decided to go with Happo as the go-to visual tests provider.

Comment (denieler) [reply docs/decisions/04-single-visual-tests-tool.md:28]: yeah, I also agree that if we remove Storybook visual tests completely - we automatically add a huge maintenance effort of covering all those use-cases by Cypress tests. And the second - we might not be sure that deployed storybook looks correct

File: docs/decisions/04-single-visual-tests-tool.md (hunk 4 above)
Comment (denieler) [reply docs/decisions/04-single-visual-tests-tool.md]: yeah, I've checked both pages and it looks like the effort pretty small - configs are just 1-2 liners looks like

File: docs/decisions/04-single-visual-tests-tool.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+## Problem
+
+Two different frameworks for visual testing are currently supported in Picasso:
+1. Puppeteer with Storybook
+2. Happo with Cypress
+
+The Puppeteer based stack has some downsides:
+- screenshots are hard to update
+- diffs inspection is not straightforward
```
Comment (denieler) [inline docs/decisions/04-single-visual-tests-tool.md:11]: idk, I never had issues with diffs inspections. I even like it more in GitHub rather than in Happo UI

Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md:11]: I should've said local inspection, rather than on Github. :thinking:

File: docs/decisions/04-single-visual-tests-tool.md (hunk 4 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md]: While the effort is small, it's still true that we can't merge the configs, and we need two separate configurations.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 8 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md]: It's possible via testing Storybook. But it still adds a layer of complexity. I've added the information in the RFC.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 7 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md]: Yes, that is the idea. I've tried clarifying.

File: docs/decisions/04-single-visual-tests-tool.md (hunk 6 above)
Comment (separatio) [reply docs/decisions/04-single-visual-tests-tool.md]: Yeah, this was confusing. I re-wrote this section.

And even if we copy-paste examples that we need testing, that's fine, as long as we _know_ what we are testing.
Right now, we are mostly assuming all examples are tested and we don't have any reporting for visual tests in Storybook, which is a downside.

**Review by OleksandrNechai — APPROVED** [review]

**Review by TomasSlama — APPROVED** [review]

File: docs/decisions/04-single-visual-tests-tool.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+This section was added to clarify the options for Picasso with the Happo provider.
+The options are:
+- pure Storybook plugin, all visual tests being done in Storybook
+- mix between Storybook and Cypress plugins, visual test being split, most likely with two separate CI jobs
+- pure Cypress plugin, all visual tests being done in Cypress
+
+## Proposal
+
+Replace the Puppeteer stack with the Happo provider plugin for Storybook, while **keeping** Cypress tests.
```
Comment (denieler) [inline docs/decisions/04-single-visual-tests-tool.md:29]:
```suggestion
Replace the Puppeteer stack with the Happo provider plugin for Storybook, while **keeping** Cypress tests with Happo too.
```

**Review by denieler — APPROVED** [review]
